plugins/connection/grpc.py

```diff
@@ -0,0 +1,292 @@
+  persistent_connect_timeout:
+    type: int
+    description:
+      - Configures, in seconds, the amount of time to wait when trying to
+        initially establish a persistent connection. If this value expires
+        before the connection to the remote device is completed, the connection
+        will fail.
+    default: 30
+    ini:
+      - section: persistent_connection
+        key: connect_timeout
+    env:
+      - name: ANSIBLE_PERSISTENT_CONNECT_TIMEOUT
+    vars:
+      - name: ansible_connect_timeout
+  persistent_command_timeout:
+    type: int
+    description:
+      - Configures, in seconds, the default timeout
+        value when awaiting a response after issuing a call to a RPC. If the RPC
+        does not return in timeout seconds, an error is generated and the connection is
+        closed.
+    default: 30
+    ini:
+      - section: persistent_connection
+        key: command_timeout
+    env:
+      - name: ANSIBLE_PERSISTENT_COMMAND_TIMEOUT
+    vars:
+      - name: ansible_command_timeout
+  persistent_log_messages:
+    type: boolean
+    description:
+      - This flag will enable logging the command executed and response received from
+        target device in the ansible log file. For this option to work the 'log_path' ansible
+        configuration option is required to be set to a file path with write access.
+      - Be sure to fully understand the security implications of enabling this
+        option as it could create a security vulnerability by logging sensitive information in log file.
+    default: False
+    ini:
+      - section: persistent_connection
+        key: log_messages
+    env:
+      - name: ANSIBLE_PERSISTENT_LOG_MESSAGES
+    vars:
+      - name: ansible_persistent_log_messages
+  grpc_type:
+    type: str
+    description: grpc type
+    default: False
+    ini:
+      - section: grpc_connection
+        key: type
+    env:
+      - name: ANSIBLE_GRPC_CONNECTION_TYPE
+    vars:
+      - name: ansible_grpc_connection_type
+"""
+
+from ansible.errors import AnsibleConnectionFailure, AnsibleError
+from ansible.plugins.connection import NetworkConnectionBase
+from importlib import import_module
+
+try:
+    from grpc import ssl_channel_credentials, secure_channel, insecure_channel
+    from grpc.beta import implementations
+    HAS_GRPC = True
+except ImportError:
+    HAS_GRPC = False
+
+try:
+    from google import protobuf
+    HAS_PROTOBUF = True
+except ImportError:
+    HAS_PROTOBUF = False
+
+
+class Connection(NetworkConnectionBase):
+    """GRPC connections"""
+
+    transport = 'grpc'
+    has_pipelining = False
+
+    def __init__(self, play_context, new_stdin, *args, **kwargs):
+        super(Connection, self).__init__(play_context, new_stdin, *args, **kwargs)
+
+        # TODO: Need to add support to make grpc connection work with non-network target host.
+        # Currently this works only with network target host.
+        if self._network_os:
```

Comment:
iiuc the use for `grpc_type` option I think it should also check it user has set the `grpc_type` configuration option with a `or` condition.